lightning/src/ln/channelmanager.rs

```diff
@@ -6122,14 +6122,50 @@ impl<Signer: Sign, M: Deref , T: Deref , K: Deref , F: Deref , L: Deref >
 	}
 
 	fn provided_node_features(&self) -> NodeFeatures {
-		NodeFeatures::known()
+		provided_node_features()
 	}
 
 	fn provided_init_features(&self, _their_init_features: &PublicKey) -> InitFeatures {
-		InitFeatures::known_channel_features()
+		provided_init_features()
 	}
 }
 
+/// Fetches the set of [`NodeFeature`] flags which are provided by or required by
+/// [`ChannelManager`].
+pub fn provided_node_features() -> NodeFeatures {
+	provided_init_features().to_context()
+}
+
+/// Fetches the set of [`InvoiceFeature`] flags which are provided by or required by
+/// [`ChannelManager`].
+pub fn provided_invoice_features() -> InvoiceFeatures {
```

Comment:
Should this be test-only?

Reply:
`InvoiceBuilder` has methods for setting individual features (or sets of features) along with any necessary data needed in the invoice for the feature. A user never passes `InvoiceFeatures` directly. This allows compile-time enforcement of correctness.

For example `InvoiceBuilder::payment_secret` sets the supplied `PaymentSecret` and the necessary feature bits. It also makes `InvoiceBuilder::basic_mpp` available.

Reply:
Looks like `test_builder_ok` caught this because `InvoiceBuilder::payment_secret` sets dependent feature `variable_length_onion`.

Reply:
Or at very least sets it optional rather than required.

Reply:
Hmm.... maybe something else is going on there. Looking.

Reply:
Ok, it just looks like `test_builder_ok` is not manually setting `variable_length_onion`. It isn't using `provided_invoice_features`.

Reply:
> Ah, right, I suppose if we ever have a feature bit in the invoice which doesn't require some additional data in the invoice we'd need this. In any case, I don't see any harm in leaving it public.

We already have that with `basic_mpp`, but it's just a separate builder method. I don't think we should allow users to pass in `InvoiceFeatures` directly. Maybe would need to set an unknown bit, but in that case we can provide a method to do so.

Reply:
> > I don't think we should allow users to pass in InvoiceFeatures directly.
> 
> Seems like something that is going to depend on the types of features. If we end up with some feature which doesn't require additional data, we'll need this.
> 
> I kinda wonder if we should have some kind of "check features are X" method - require that users pass the `ChannelManager`'s invoice features set to the `InvoiceBuilder` so that the `InvoiceBuilder` can check that the required features have all been filled in.

`InvoiceBuilder::build_signed` returns `Result<Invoice, CreationError>`, so it could perform checks against any features, provided either at builder construction time or at invoice building time using a dedicated method.